### PR TITLE
Add quadrature on the square and cube

### PR DESCRIPTION
**Pull request by alexfikl**
This adds the quadrature rules from [Witherden and Vincent](https://doi.org/10.1016/j.camwa.2015.03.017).

I also added an extractor for the rules from [Festa and Sommariva](https://doi.org/10.1016/j.cam.2012.05.021) that can be found [here](https://www.math.unipd.it/~alvise/sets.html) before realizing those are GPL licensed. They're also just for the square (not the cube), but have orders up to 55 while WV only goes to 20.

### CODE REVIEW

File: modepy/tools.py [hunk 1]
```diff
@@ -101,6 +101,13 @@ def simplex_integral(self):
                 * reduce(mul, (factorial(alpha) for alpha in self.exponents))
                 / factorial(len(self.exponents)+sum(self.exponents)))
 
+    def hypercube_integral(self):
+        """Integral over the unit hypercube."""
```
Comment (inducer) [inline modepy/tools.py:105]:
```suggestion
        """Integral over :math:`[0,1]^n`."""
```

File: modepy/quadrature/witherden_vincent_quad_data.py [hunk 2]
```diff
@@ -0,0 +1,2106 @@
+# GENERATED by modepy/contrib/weights-from-witherden-vincent.py
```
Comment (inducer) [inline modepy/quadrature/witherden_vincent_quad_data.py:1]: Maybe generate a copyright/license statement here?

File: contrib/weights-from-festa-sommariva.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    return result
+
+
+square_data = process_rules(%s)"""
+
+
+def generate_festa_sommariva_quadrature_rules(outfile, pointset="set_amr_square"):
+    filename = f"{pointset}.m"
+    download_from_web_if_not_present(
+            url=f"https://www.math.unipd.it/~alvise/POINTSETS/{filename}",
```
Comment (inducer) [inline contrib/weights-from-festa-sommariva.py:59]: Ask https://archive.org to archive this and use the resulting URL.

File: modepy/tools.py (hunk 1 above)
Comment (inducer) [reply modepy/tools.py:105]: If you're feeling kind, maybe also define the "unit simplex" above.

**Review by inducer — COMMENTED** [review]
Looks great. Just a couple wrinkles.

File: contrib/weights-from-festa-sommariva.py (hunk 3 above)
Comment (alexfikl) [reply contrib/weights-from-festa-sommariva.py]: Fixed in 1e73044 (hopefully that's what you meant)

File: modepy/tools.py (hunk 1 above)
Comment (alexfikl) [reply modepy/tools.py]: Fixed in a4363b0.

Is there a more compact expression for the simplex?

File: modepy/quadrature/witherden_vincent_quad_data.py (hunk 2 above)
Comment (alexfikl) [reply modepy/quadrature/witherden_vincent_quad_data.py:1]: For the original authors? How would that look like?

File: modepy/tools.py (hunk 1 above)
Comment (inducer) [reply modepy/tools.py]: Thx! Not that I know of.

File: modepy/quadrature/witherden_vincent_quad_data.py (hunk 2 above)
Comment (inducer) [reply modepy/quadrature/witherden_vincent_quad_data.py:1]:
```py
__copyright__ = "(C) Witherden and company"
__license__ = """
The glorious text of the applicable CC license
"""
```
?

Comment (alexfikl) [reply modepy/quadrature/witherden_vincent_quad_data.py:1]: How does 73d718f look?  There doesn't seem to be a standard header for CC licenses.

**Comment by inducer** [comment]
Thanks!

**Pull request merged**